cmd/tsdbrelay/denormalization.go

```diff
@@ -0,0 +1,65 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"bosun.org/opentsdb"
+)
+
+type DenormalizationRule struct {
+	Metric   string
+	TagNames []string
+}
+
+func (d *DenormalizationRule) String() string {
+	inputTags, outputTags := "", ""
+	val := 'a'
+	for i, tagk := range d.TagNames {
+		if i != 0 {
+			inputTags += ","
+			outputTags += "."
+		}
+		inputTags += fmt.Sprintf("%s=%s", tagk, string(val))
+		outputTags += fmt.Sprintf("%s", string(val))
+		val++
+	}
+	return fmt.Sprintf("%s{%s} -> __%s.%s", d.Metric, inputTags, outputTags, d.Metric)
+}
+
+var denormalizationRules map[string]*DenormalizationRule
+
+func parseDenormalizationRules(config string) (map[string]*DenormalizationRule, error) {
+	m := make(map[string]*DenormalizationRule)
+	rules := strings.Split(config, ",")
+	for _, r := range rules {
+		parts := strings.Split(r, "__")
+		if len(parts) < 2 {
+			return nil, fmt.Errorf("Denormalization rules must have at least one tag name specified.")
+		}
+		rule := &DenormalizationRule{Metric: parts[0]}
+		for i := 1; i < len(parts); i++ {
+			rule.TagNames = append(rule.TagNames, parts[i])
+		}
+		log.Println("Denormalizing", rule)
+		m[rule.Metric] = rule
+	}
+	return m, nil
+}
+
+func (d *DenormalizationRule) Translate(dp *opentsdb.DataPoint) (*opentsdb.DataPoint, error) {
```

Comment:
Since this now modifies dp in place, I think the return signature should just be `error`. In our code we have `newDp := d.Translate(...)`, which implies a new dp is created, which is dangerous.